bin/varnishd/cache/cache_hash.c

```diff
@@ -355,6 +355,63 @@ hsh_insert_busyobj(const struct worker *wrk, struct objhead *oh)
 	return (oc);
 }
 
+/*---------------------------------------------------------------------
+ */
+
+static void
+hsh_waitlist(struct worker *wrk, struct req *req, struct objhead *oh)
+{
+
+	CHECK_OBJ_NOTNULL(wrk, WORKER_MAGIC);
+	CHECK_OBJ_NOTNULL(req, REQ_MAGIC);
+	CHECK_OBJ_NOTNULL(oh, OBJHEAD_MAGIC);
+	Lck_AssertHeld(&oh->mtx);
+	AZ(req->hash_ignore_busy);
+	AZ(req->hash_objhead);
+
+	DSLb(DBG_WAITINGLIST, "on waiting list <%p>", oh);
+
+	/*
+	 * The objhead reference transfers to the sess, we get it
+	 * back when the sess comes off the waiting list and
+	 * calls us again
+	 */
+	req->hash_objhead = oh;
+	req->wrk = NULL;
+	req->waitinglist = 1;
+
+	if (req->top->topreq->transport->waitlist != NULL)
+		req->top->topreq->transport->waitlist(wrk, req);
```

Comment:
In the commit where it is introduced, there is a comment stating that it is only for `req_top`'s transport. The reason for the lack of symmetry is that the "current" transport action is the same for all requests: go into the waiting list.

Without this patch series there is already a reembark asymmetry.

What we have done downstream is less idea in my opinion, because we made each transport responsible for checking whether there is a top `waitlist` callback to signal instead of centralizing this check like in the current iteration.

We could have a distinction between the sole `reembark` callback and two separate `top_waitlist` and `top_reembark`. I didn't because I felt it was too heavy-handed.

Reply:
> We could have a distinction between the sole `reembark` callback and two separate `top_waitlist` and `top_reembark`. I didn't because I felt it was too heavy-handed.

I think I prefer that option.

Reply:
Done.